Ensure the spread of pods on different nodes

aledbf opened the pull request:
## Description

Hint we prefer to spread pods across the available nodes in the cluster

https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/

## How to test
- Integration tests should pass
 
## Release Notes
```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

WVerlaek commented:
/werft with-werft with-integration-tests=workspace with-large-vm

WVerlaek commented:
/werft with-integration-tests=workspace with-large-vm

WVerlaek reviewed (approved):
lgtm from workspace side

akosyakov commented on install/installer/pkg/components/blobserve/deployment.go:
```diff
@@ -77,9 +77,10 @@ func deployment(ctx *common.RenderContext) ([]runtime.Object, error) {
-						Affinity:           common.NodeAffinity(cluster.AffinityLabelIDE),
-						ServiceAccountName: Component,
-						EnableServiceLinks: pointer.Bool(false),
+						Affinity:                  cluster.WithNodeAffinityHostnameAntiAffinity(Component, cluster.AffinityLabelIDE),
```
@iQQBot Could you have a look please? 🙏

aledbf commented:
> 1. we _try_ to distribute pods evenly between nodes

Correct

> 2. if that does not add up (because we have, say, 7 pods, and 4 nodes), we end up with a 2, 2, 2, 1 distribution, because of  `MaxSkew: 1` in combination with `whenUnsatisfiable: ScheduleAnyway`?

Yes. if the maxSkew cannot be fulfilled, it will return to what the k8s scheduler decides and schedule the pod/s using binpacking.

aledbf commented:
/werft run

aledbf commented:
> That is a lot of code 🙃

Please filter the code to only see the `.go` changes.

aledbf commented:
/werft run with-clean-slate-deployment=true

aledbf commented on install/installer/pkg/components/migrations/job.go:
```diff
@@ -34,10 +34,11 @@ func job(ctx *common.RenderContext) ([]runtime.Object, error) {
-					RestartPolicy:      corev1.RestartPolicyNever,
-					ServiceAccountName: Component,
-					EnableServiceLinks: pointer.Bool(false),
+					Affinity:                  cluster.WithNodeAffinityHostnameAntiAffinity(Component, cluster.AffinityLabelMeta),
```
fixed

aledbf commented on install/installer/pkg/components/openvsx-proxy/statefulset.go:
```diff
@@ -93,11 +93,12 @@ func statefulset(ctx *common.RenderContext) ([]runtime.Object, error) {
 				},
 				Spec: v1.PodSpec{
-					Affinity:                      common.NodeAffinity(cluster.AffinityLabelIDE),
+					Affinity:                      cluster.WithNodeAffinityHostnameAntiAffinity(Component, cluster.AffinityLabelIDE),
```
Reverted

aledbf commented:
> some question:
> 
> * We removed most of the network policy, but why?
> * Some components of the service add metrics ports, some don't.

This was a mistake. Fixed.

> * There is no need to use AntiAffinity for job type tasks.

Fixed

geropl reviewed (approved):
LGTM
